### PR TITLE
Expose targetline colors to yaml

### LINKED ISSUE #18467: Change lines color to make orders easily identifiable
## Request motivation

Currently, move orders are identified by a green line, while both the A-move and the attack action are represented as red lines -- it's not exactly the same red, but it's very hard do distinguish in-game.

OraLines

## Proposed solution

I would suggest to change attack move color to blue.

## Side effects

None :smile: 

## Alternatives

Any other distinguishable color would work.

### PR DESCRIPTION
**Pull request by tovl**
Fixes #18467

- Exposes all targetline colors to yaml so that they can be changed by mods.
- ~Changes the default targetline colors to be more distinguishable.~
- ~Gives all heal actions in standard mods the same unique color.~
- Removed target lines from scripted activities.

### CODE REVIEW

File: OpenRA.Mods.Cnc/Activities/LayMines.cs
Comment (Mailaender) [inline OpenRA.Mods.Cnc/Activities/LayMines.cs:125]: Maybe fetch that from `MineLayer` instead?

**Review by Mailaender — DISMISSED** [review]
Excellent idea. Works as promised.

**Comment by pchote** [comment]
I do not have time to test or look at the code, but want to again raise my objections (which I had already explained in previous discussions) against making heal blue.

**Comment by tovl** [comment]
You never did raise any objections to making heal blue. You just stated you didn't like it. Anything else was already thoroughly rebutted in those very discussions. You yourself admitted back then that your feelings on the matter where just a knee-jerk reaction, without a logical explanation behind it. 

I didn't push the matter further because I thought going for orange-red was a good enough alternative for readability anyway. Since it now turns out that some people do find orange-red not distinguishable enough, it makes sense to separate the colors further which can only be done by going more towards yellow. This would then necessitate  providing a different color for heal actions.

As well, If you would have actually tested _or_ read the code, you would have known that this isn't anywhere near the particular shade of blue that you found so 'painful' to read. It is in fact much more towards the Cyan side, which is something you yourself even suggested at one point!

**Comment by deleted-user-1** [comment]
I'm the drama contributor already and there can only be one of us! /sarc

On a more serious note, I would welcome if line-colors were configurable at least in the settings yaml (does not need be in the UI) because of colorblindness...

**Comment by JustSomeDev1** [comment]
That's the color 
#00bfff

My 2 cents: I personally think it's fine, the only concern may be about its visibility in a sea map but since it's reserved to heal and repair, that shouldn't be a big issue.

**Comment by Mailaender** [comment]
If it is configurable in the trait settings (which is my preference as it is more per mod thing taking tileset backgrounds and UI colors into account) then we can quickly iterate on that and change the colors afterwards easily. I see zero potential for drama.

**Comment by deleted-user-1** [comment]
Yeah, well, guess I'm bad at jokes. Configurable colors would still be neat for accessibility reasons but I raised that idea more to offer a compromise for what I probably misinterpreted as bikeshedding over colors than as a "review request".

**Comment by deleted-user-1** [comment]
FYI
> I do not have time to test or look at the code, but want to again raise my objections (which I had already explained in previous discussions) against making heal blue.

Potentially referring to https://github.com/OpenRA/OpenRA/pull/16549#issuecomment-494993986.

>  that this isn't anywhere near the particular shade of blue that you found so 'painful' to read. It is in fact much more towards the Cyan side, which is something you yourself even suggested at one point!

Potentially referring to https://github.com/OpenRA/OpenRA/issues/16616#issuecomment-497106231

File: OpenRA.Mods.Cnc/Activities/LayMines.cs
Comment (abcdefg30) [inline OpenRA.Mods.Cnc/Activities/LayMines.cs]: Crashing here when neither `mobile` nor `aircraft` are used as `IMove` by the actor looks rather odd to me. What do you think about having `IMoveInfo` define a `GetTargetLineColor` method which is overwritten/implemented by `MobileInfo` and `AircraftInfo` and called here instead of "hardcoding" specific traits?

File: OpenRA.Mods.Common/Activities/Resupply.cs
Comment (abcdefg30) [inline OpenRA.Mods.Common/Activities/Resupply.cs]: Same could then be used here.

File: OpenRA.Mods.Cnc/Activities/LayMines.cs
Comment (tovl) [reply OpenRA.Mods.Cnc/Activities/LayMines.cs]: Good idea! fixed.

File: OpenRA.Mods.Common/Traits/Repairable.cs
Comment (abcdefg30) [inline OpenRA.Mods.Common/Traits/Repairable.cs]: Why is this adding a using without using any new code?

File: OpenRA.Mods.Cnc/Traits/TDGunboat.cs
Comment (abcdefg30) [inline OpenRA.Mods.Cnc/Traits/TDGunboat.cs]: Style nit: Comments should go on their own line. I also don't think we need a comment here, but don't care really either way.

**Comment by tovl** [comment]
Applied fixups.

File: OpenRA.Mods.Cnc/Activities/Infiltrate.cs
Comment (pchote) [inline OpenRA.Mods.Cnc/Activities/Infiltrate.cs]: Style nit: call this `targetLineColor` to match `Enter`.

File: OpenRA.Game/Traits/TraitsInterfaces.cs
Comment (pchote) [inline OpenRA.Game/Traits/TraitsInterfaces.cs:512]: Would a property be better?

File: OpenRA.Mods.Cnc/Scripting/Properties/InfiltrateProperties.cs
Comment (pchote) [inline OpenRA.Mods.Cnc/Scripting/Properties/InfiltrateProperties.cs:41]: This and the other Lua API changes will prevent the target lines from being shown. This is a non-trivial behaviour change that wasn't mentioned in the PR description!

File: OpenRA.Mods.Cnc/Traits/PortableChrono.cs
Comment (pchote) [inline OpenRA.Mods.Cnc/Traits/PortableChrono.cs]: This isn't a target line! It is also being unhardcoded by #17919, so it would be best to drop it from here.

File: OpenRA.Mods.Cnc/Traits/TDGunboat.cs
Comment (pchote) [inline OpenRA.Mods.Cnc/Traits/TDGunboat.cs:35]: We should extract this into a `TargetLineColor` field for consistency.

File: OpenRA.Mods.Common/Activities/CaptureActor.cs
Comment (pchote) [inline OpenRA.Mods.Common/Activities/CaptureActor.cs]: `targetLineColor`

File: OpenRA.Mods.Common/Activities/Demolish.cs
Comment (pchote) [inline OpenRA.Mods.Common/Activities/Demolish.cs]: `targetLineColor`

File: OpenRA.Mods.Common/Activities/DonateCash.cs
Comment (pchote) [inline OpenRA.Mods.Common/Activities/DonateCash.cs]: `targetLineColor`

File: OpenRA.Mods.Common/Activities/DonateExperience.cs
Comment (pchote) [inline OpenRA.Mods.Common/Activities/DonateExperience.cs]: `targetLineColor`

File: OpenRA.Mods.Common/Activities/RepairBridge.cs
Comment (pchote) [inline OpenRA.Mods.Common/Activities/RepairBridge.cs]: `targetLineColor`

File: OpenRA.Mods.Common/Activities/RideTransport.cs
Comment (pchote) [inline OpenRA.Mods.Common/Activities/RideTransport.cs]: `targetLineColor`

File: OpenRA.Mods.Cnc/Scripting/Properties/InfiltrateProperties.cs
Comment (pchote) [reply OpenRA.Mods.Cnc/Scripting/Properties/InfiltrateProperties.cs:41]: I understand your logic here based on previous discussions: automatic behaviours should not have target lines - but i'm not sure this is correct here, because player-controlled units may have commands scripted and these *should* display when the user presses the shift key.

File: OpenRA.Mods.Common/Traits/EntersTunnels.cs
Comment (pchote) [inline OpenRA.Mods.Common/Traits/EntersTunnels.cs]: `moveInfo` and `moveInfo.GetTargetLIneColor()` below.

**Comment by pchote** [comment]
I'm still not a fan of the colour changes, but i'm willing to let that go provided:
* we get feedback from a wider set of players (polling on discord would be enough) that this is wanted
* the colour groupings still make sense (e.g. shouldn't carryalls now also be blue? this is a support action, not attackmove/guard!)

**Comment by deleted-user-1** [comment]
> polling on discord would be enough

Forum* would be more representative than asking the <5 opinion leaders on discord. Discord is a formalized opionion hierarchy. Please don't classify whatever you get there as feedback from a wider set of players.

* I would utilize (in)gamenews for this

**Comment by pchote** [comment]
> Forum* would be more representative

Ok, sure.

**Comment by pchote** [comment]
Any progress here?

At the least, it would be good to get the unhardcoding in for the next release even if we think the colour changes are too much right now.

**Comment by pchote** [comment]
Ping @tovl. If you're not able/wanting to continue with this in the short term i'll file a new PR that focuses on just exposing the hardcoded colors to yaml.

**Comment by tovl** [comment]
I'm perfectly able and willing to continue with this. The question is: are you able and willing to let me?

Dropping the colour changes would mean this PR fails to fix the one issue this PR was meant to fix. That is a reasonable concern, is it not? I would've been glad to just bury this topic after the friction it caused the last time it came up, if it wasn't for that issue being posted.

**Comment by pchote** [comment]
Yeah, like I said above i'm happy to let my personal objection go provided we can demonstrate that there really is a desire from at least some cross section of active players (more than just one or two individuals) to do so and then make sure that the new scheme follows internally consistent rules for grouping related types of orders.

My motivation for splitting this into two is that it is nearly impossible to get useful feedback from players unless the changes can be applied to custom maps on the current active release. This obviously isn't possible right now, but we can make sure it is possible for the next.

**Comment by tovl** [comment]
Okay, that sounds reasonable. I'll split the PR and rebase soonish.

File: OpenRA.Game/Traits/TraitsInterfaces.cs
Comment (tovl) [reply OpenRA.Game/Traits/TraitsInterfaces.cs:512]: This is consistent with how similar cases are handled throughout the code, e.g. in `AircraftInfo`.

File: OpenRA.Mods.Cnc/Traits/TDGunboat.cs
Comment (tovl) [reply OpenRA.Mods.Cnc/Traits/TDGunboat.cs:35]: I'd rather not expose a parameter that would never be actually used.

File: OpenRA.Mods.Cnc/Scripting/Properties/InfiltrateProperties.cs
Comment (tovl) [reply OpenRA.Mods.Cnc/Scripting/Properties/InfiltrateProperties.cs:41]: I'm not sure they should. Scripted behaviours are things that are outside of the player's control (in the general sense) and I think it would be illogical and immersion breaking for the player to have foreknowledge of queued activities that where queued outside of the player's control.

Besides, wouldn't player-control over a scripted unit be inherently contradictory anyway? As far as I remember, the original games would always temporarily remove a unit from the player's control when it needed to perform a scripted action.

**Comment by tovl** [comment]
Rebased and split of the color changes.

File: OpenRA.Mods.Common/Scripting/Properties/DemolitionProperties.cs
Comment (abcdefg30) [inline OpenRA.Mods.Common/Scripting/Properties/DemolitionProperties.cs:36]: Why isn't this using the color from the info?

File: OpenRA.Mods.Common/Traits/EntersTunnels.cs
Comment (abcdefg30) [inline OpenRA.Mods.Common/Traits/EntersTunnels.cs]: This is unused.

File: OpenRA.Mods.Cnc/Scripting/Properties/InfiltrateProperties.cs
Comment (pchote) [reply OpenRA.Mods.Cnc/Scripting/Properties/InfiltrateProperties.cs:41]: Ok. Considering that this is currently a hypothetical situation I am okay with letting it be until/if it ever becomes an issue in practice. Silently changing the behaviour is still a problem though, can we please at least mention this in the extended commit message, and ideally split it into its own commit to improve visibility in the git history and blame?

File: OpenRA.Mods.Cnc/Traits/TDGunboat.cs
Comment (pchote) [reply OpenRA.Mods.Cnc/Traits/TDGunboat.cs:35]: Ok.

File: OpenRA.Game/Traits/TraitsInterfaces.cs
Comment (pchote) [reply OpenRA.Game/Traits/TraitsInterfaces.cs:512]: Ok.

File: OpenRA.Mods.Common/Scripting/Properties/DemolitionProperties.cs
Comment (abcdefg30) [reply OpenRA.Mods.Common/Scripting/Properties/DemolitionProperties.cs:36]: Actually I just saw the discussion above. This should probably get a comment so nobody "fixes it" by accident.

Comment (tovl) [reply OpenRA.Mods.Common/Scripting/Properties/DemolitionProperties.cs:35]: See discussion above.

**Comment by tovl** [comment]
Resolved remaining fixups. Added extra comments in the code and the commit message regarding scripted activities.

**Review by abcdefg30 — DISMISSED** [review]
Works as advertised. Fwiw, I'm still not a fan of not showing target lines, but that is a different issue.

File: OpenRA.Mods.Common/Traits/EntersTunnels.cs
Comment (pchote) [inline OpenRA.Mods.Common/Traits/EntersTunnels.cs]: This should be `Color.Green`.

**Review by pchote — COMMENTED** [review]
LGTM, just one color change that wasn't caught. I will be away for the next week or so, so please merge this after fixup without waiting for me to reconfirm my review (assuming it is just the color change).

**Comment by pchote** [comment]
Pushed to rebase and fix my review nit above.

**Review by pchote — APPROVED** [review]

**Comment by pchote** [comment]
https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/2fe27b097063a7b9117149d1a9ea57cb7aaf5198

**Pull request merged**
